### DISCUSSION TITLE
Publishing in Module 4 Assignment - `_queue.Empty`

### DISCUSSION THREAD

**Discussion by chasedamagician** [AccelerationConsortium/ac-microcourses#19] [Hello World] [answered]
Hello, I am currently working on the Module 4 Assignmnet and I am having some difficulties publishing the dictionary and results. Whenever I test the programs I get the following errors:

```
"FAILED microcontroller_client_test.py::test_send_and_receive - _queue.Empty: Did not receive any data on topic adamant-ferret/as7341 after publishing to adamant-ferret/neopixel. 

FAILED orchestrator_client_test.py::test_orchestrator_client - Exception: results.json and payload_dicts.json not found within 30 s. Number of commands received so far: 0."

```

currently I am publishing the payload_dictionary and results as .json strings, I believe the issue is due to how I publish the strings:

```
"results_str = json.dumps(payload_dict)
   client.publish(command_topic, results_str, qos = 2, retain = False) "

```
and 

```
"dict_str = json.dumps(formatted_dict)
 client.publish(sensor_data_topic, dict_str, qos=2, retain=False)"
```

Thank you for your time.

**Comment by sgbaird** [comment]
I think you also mentioned that you are seeing the messages on the HiveMQ broker. Do you see any different output on the microcontroller when you're running it?

**Reply by chasedamagician** [reply]
Yes, on the microcontroller for the experiment_id  "6fe7fa1f", I see:

`('adamant-ferret/neopixel', '{"command": {"R": 15, "G": 82, "B": 186}, "experiment_id": "6fe7fa1f"}', 'False')`

While on the Broker the output I see from the the "course_id/as7341"(sensor_data_topic) topic is:

`
  | {"command": {"R": 15, "G": 82, "B": 186}, "experiment_id": "6fe7fa1f", "sensor_data": {"ch410": 185.3, "ch440": 221.8, "ch470": 202.5, "ch510": 183.2, "ch550": 163.9, "ch583": 111.80000000000001, "ch620": 64.9, "ch670": 31.400000000000002}}`

additionally for the "course_id/neopixel"(command_topic) topic I get this output:

`{"command": {"R": 15, "G": 82, "B": 186}, "experiment_id": "6fe7fa1f"}`

**Reply by sgbaird** [reply]
Can you try running `orchestrator_client.py` directly? (i.e., not with pytest - just F5 / play button). Also, can you include the GitHub link to your `orchestrator_client.py` file? (after committing your most recent changes)

**Reply by chasedamagician** [reply]
When I do so, I get the following error:
` File "/workspaces/4-hardware-software-communication-chasedamagician/orchestrator_client.py", line 123, in <module>
    results_dict = run_experiment(client, queue, neopixel_topic, payload_dict)
  File "/workspaces/4-hardware-software-communication-chasedamagician/orchestrator_client.py", line 95, in run_experiment
    raise Empty(`

and when I check the broker I do not get an output from course_id/as7341 so I suspect that the issue is coming from the microcontroller_client program, but I am not completely sure, do you have any suggestions? Thank you

**Reply by sgbaird** [reply]
I'm not seeing any implementation for `messages`. Are these the most recent changes?

https://github.com/ACC-HelloWorld/4-hardware-software-communication-chasedamagician/blob/main/microcontroller_client.py#L112

**Reply by sgbaird** [reply]
I see your recent changes now. Thanks! Can you try this with the default HiveMQ credentials? (the public, test broker). Nothing jumps out at me that's immediately obvious.

EDIT: @chasedamagician could you comment how you resolved the issue?

**Reply by chasedamagician** [reply, marked as answer]
Yes, one of the main issues was that I was not careful of what QOS I used for the sensor_data_topic used by the microcontroller, I was using QOS 2 but for the Raspberry Pico I should have been using QOS 0 or 1, I also used the await function when I published the dictionary so the program would wait for the publish command before proceeding.
